yt/utilities/lib/image_utilities.pyx

```diff
@@ -36,6 +41,370 @@ def add_points_to_greyscale_image(
+    cdef np.uint8_t[:, :, ::1] stamp_mask = stamp_mask_arr
+
+    # Precompute the mip
+    for depth in range(max_depth):
+        if depth == 0:
+            stamp[0, 0, 0] = 1
+            continue
+        direct_integrate_cube(
+            o,
+            a,
+            b,
+            c,
+            stamp_arr[depth, :, :],
+            stamp_mask_arr[depth, :, :],
+            min(1 << depth, Nsx),
+            min(1 << depth, Nsy),
+        )
+
+        stamp_arr[depth] /= np.sum(stamp[depth])
+
+    # Iterate over all cells, applying the stamp
+    cdef np.float64_t x, y, dx
+    cdef np.float64_t[:, ::1] rotation_view = np.ascontiguousarray(rotation)
+
+    cdef np.float64_t w, q, cell_max_width, sq3
+
+    sq3 = sqrt(3.)
+
+    # Local buffers
+    cdef np.float64_t **lbuffer
+    cdef np.float64_t **lbuffer_weight
+
+    cdef int num_particles = len(Xp)
+
+    with nogil, parallel():
+        lbuffer = <np.float64_t**> malloc(sizeof(np.float64_t*) * Nx)
+        lbuffer_weight = <np.float64_t**> malloc(sizeof(np.float64_t*) * Nx)
+        for j in range(Nx):
+            lbuffer[j] = <np.float64_t*> malloc(sizeof(np.float64_t) * Ny)
+            lbuffer_weight[j] = <np.float64_t*> malloc(sizeof(np.float64_t) * Ny)
+            for k in range(Ny):
+                lbuffer[j][k] = 0
+                lbuffer_weight[j][k] = 0
+
+        for i in prange(num_particles, schedule="static"):
+            dx = dXp[i]
+            w = weight[i]
+            q = qty[i]
+            cell_max_width = dx * sq3
+            x = (
+                rotation_view[0, 0] * Xp[i, 0] +
+                rotation_view[0, 1] * Xp[i, 1] +
+                rotation_view[0, 2] * Xp[i, 2]
+            ) + 0.5
+            y = (
+                rotation_view[1, 0] * Xp[i, 0] +
+                rotation_view[1, 1] * Xp[i, 1] +
+                rotation_view[1, 2] * Xp[i, 2]
+            ) + 0.5
+
+            _add_cell_to_image_offaxis(
+                dx,
+                w,
+                q,
+                cell_max_width,
+                x,
+                y,
+                Nx,
+                Ny,
+                Nsx,
+                Nsy,
+                lbuffer,
+                lbuffer_weight,
+                max_depth,
+                stamp,
+                stamp_mask
+            )
+
+        # Copy back data in main buffer
+        with gil:
+            for j in range(Nx):
+                for k in range(Ny):
+                    buffer[j, k] += lbuffer[j][k]
+                    buffer_weight[j, k] += lbuffer_weight[j][k]
+        # Free memory
+        for j in range(Nx):
+            free(lbuffer[j])
+            free(lbuffer_weight[j])
+        free(lbuffer)
+        free(lbuffer_weight)
+
+
+@cython.boundscheck(False)
+cdef inline np.float64_t det2d(const np.float64_t[::1] a, const np.float64_t[::1] b) noexcept nogil:
+    return a[0] * b[1] - a[1] * b[0]
+
+@cython.cdivision(True)
+cdef bint check_in_parallelogram(
```

Comment:
boundscheck and initializedcheck might add speed

Reply:
done!